### PR TITLE
PUBDEV-8607 Improve UpliftDRF demo

### PR DESCRIPTION
**Pull request by maurever**
JIRA: https://h2oai.atlassian.net/browse/PUBDEV-8607

### CODE REVIEW

**Comment by maurever** [comment]
The demo is ready for review. There is only one problem - the full-size Criteo dataset is not synced in the smalldata. There is a possibility to load it from a URL using the pandas library, but it is very time-consuming. I would maybe prefer to add the dataset into the bigdata folder, WDYT @valenad1, @michalkurka?

File: h2o-py/demos/uplift_drf_demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,2844 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## H2O Implementation (Major release 3.36)\n",
+    "\n",
+    "The H2O-3 implementation of Uplift DRF is based on DRF because the principle of training is similar to DRF. It is tree based uplift algorithm. Uplift DRF generates a forest of classification uplift trees, rather than a single classification tree. Each of these trees is a weak learner built on a subset of rows and columns. More trees will reduce the variance. Classification takes the average prediction over all of their trees to make a final prediction. \n",
+    "\n",
+    "Currently, in H2O-3 only binomial trees are supported, as well as the uplift curve metric and Area Under Uplift curve (AUUC) metric and the Qini coefficient. We are working on adding also regression trees and more metrics, for example, normalized AUUC, and more. "
```
Comment (valenad1) [inline h2o-py/demos/uplift_drf_demo.ipynb:99]: Are you sure with Qini coefficient? I remember you refactor it to Qini metric? What is the difference actually?

Comment (valenad1) [inline h2o-py/demos/uplift_drf_demo.ipynb:99]: We already have a normalized AUUC 👍

Comment (valenad1) [reply h2o-py/demos/uplift_drf_demo.ipynb:99]: Do we have something more on a list?

File: h2o-py/demos/uplift_drf_demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,2844 @@
...
+    "\n",
+    "Where:\n",
+    "\n",
+    "- **T** how many observations are in the treatment group (how many data rows in the bin have treatment_column label == 1)\n",
+    "- **C** how many observations are in the control group (how many data rows in the bin have treatment_column label == 0)\n",
+    "- **TY1** how many observations are in the treatment group and respond to the offer (how many data rows in the bin have treatment_column label == 1 and response_column label == 1)\n",
+    "- **CY1** how many observations are in the control group and respond to the offer (how many data rows in the bin have treatment_column label == 0 and response_column label == 1)\n",
+    "\n",
+    "\n",
+    "The resulting AUUC value is not normalized, so the result could be a positive number, but also a negative number. A higher number means better model. \n",
```
Comment (valenad1) [inline h2o-py/demos/uplift_drf_demo.ipynb:1187]: They are normalized now

File: h2o-py/demos/uplift_drf_demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,2844 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Scoring histrory and importance of number of trees\n",
+    "\n",
+    "To speed up the calculation of AUUC, the predictions are binned into quantile histograms. To calculate precision AUUC the more bins the better. The more trees usually produce more various predictions and then the algorithm creates histograms with more bins. So the algorithm needs more iterations to get meaningful AUUC results. \n",
+    "You can see in the scoring history table the number of bins as well as the result AUUC. There is also Qini coefficient parameter, which reflects the number of bins and then is a better pointer of the model improvement. In the picture below you can see the algorithm stabilized after building 6 trees. But it depends on data and model settings on how many trees are necessary."
```
Comment (valenad1) [inline h2o-py/demos/uplift_drf_demo.ipynb:1500]: > In the picture below you can see the algorithm stabilized after building 6 trees.

There is 0, 1, 10, 20 trees in the picture, can we show 6 trees? Or say that is stabilized after 10 Trees?

**Comment by maurever** [comment]
> Very nice demo, thank you!
> 
> Just a few points to assure:
> 
>     * The picture with Decision Tree and Uplift decision tree is ours? Do we need to cite anybody?
> 
>     * I miss some conclusion for `Comparasion Tree-based approach and Generalized Weighed Uplift` or for the whole blog at least.
> 
>     * I miss references at the end to the paper, Kaggle,..
> 
> 
> Please let me now if you can implement the suggestions or I will handle that since you are already on a leave +1

@valenad1, thanks for great review! Answers to your questions:

1. I made the picture of the trees.
2. Fixed.
3. References added. The Kaggle notebook is cited in every code block where I used something from the notebook.

File: h2o-py/demos/uplift_drf_demo.ipynb (hunk 2 above)
Comment (valenad1) [reply h2o-py/demos/uplift_drf_demo.ipynb:1422]: Here I would be more specific. I didn't get that you are talking about numbers comes from Formula up there but I connect it to the table below where is some normalized AUUC.

Can we say something like:

The resulting AUUC value is:

-  Not normalized.
- The result could be a positive or negative number. 
- Higher number means better model. 

More information about normalization is in **Normalized AUUC** section.

File: h2o-py/demos/uplift_drf_demo.ipynb [hunk 4]
```diff
@@ -63,7 +63,7 @@
     "\n",
     "In a tree node the result value for a split is sum: $metric(P, Q) + metric(1-P, 1-Q)$. \n",
     "\n",
-    "For the split gain value, the result within the node is normalized using a gini coefficient (Eclidean or ChiSquared) or entropy (KL) for each distribution before and after the split.\n",
+    "For the split gain value, the result within the node is normalized using a Gini coefficient (Eclidean or ChiSquared) or entropy (KL) for each distribution before and after the split.\n",
```
Comment (valenad1) [inline h2o-py/demos/uplift_drf_demo.ipynb:66]: Here I don't really understand. 

Are you sure with G? Or it should be Q?

Comment (valenad1) [reply h2o-py/demos/uplift_drf_demo.ipynb:66]: Yes, it's Q in the docs.. I would kill for this type of naming :D

Comment (valenad1) [reply h2o-py/demos/uplift_drf_demo.ipynb:66]: Maybe it should be Qini value?

**Comment by valenad1** [comment]
@maurever Can you explain to me what is the difference between Qini coefficient and Qini value please? Fist I thought we just rename the same thing, now I see it is a different metric but when I google Qini coefficient I got

`𝑄= Qini Curve Area−Randomized Qini Area`

which is our Qini value... The way I understand it now is that:

Qini value = Qini Curve Area−Randomized Qini Area = AECU with metric Qini

What is Qini coefficient?

**Comment by maurever** [comment]
> @maurever Can you explain to me what is the difference between Qini coefficient and Qini value please? Fist I thought we just rename the same thing, now I see it is a different metric but when I google Qini coefficient I got
> 
> `𝑄= Qini Curve Area−Randomized Qini Area`
> 
> which is our Qini value... The way I understand it now is that:
> 
> Qini value = Qini Curve Area−Randomized Qini Area = AECU with metric Qini
> 
> What is Qini coefficient?

@valenad1,

from this [source](https://www.stochasticsolutions.com/pdf/kdd2011late.pdf) I understand that Qini coefficient is something like the Gini coefficient for binomial cases. We are not able to compute the Gini coefficient due to uplift curve limitation, so the Qini coefficient is introduced. 

See picture 8 and 10: 
Screenshot from 2022-04-26 20-02-21

IMO: Qini value is in picture 8 and the Qini coefficient is then `Qini value / Best possible uplift arrea`

**Comment by valenad1** [comment]
All right, now I understand what you mean.

**Comment by valenad1** [comment]
It think the demo can go to zumbo.

**Review by valenad1 — APPROVED** [review]
Thank you!

**Review by michalkurka — APPROVED** [review]

**Pull request merged**